ultra/tests/test_replay_buffer.py

```diff
@@ -0,0 +1,341 @@
+
+import numpy as np
+
+import ultra.adapters as adapters
+from ultra.baselines.common.replay_buffer import ReplayBuffer
+
+
+class ReplayBufferTest(unittest.TestCase):
+    def test_image_replay_buffer(self):
+        TRANSITIONS = 1024  # The number of transitions to save in the replay buffer.
+        STACK_SIZE = 4  # The stack size of the images.
+        ACTION_SIZE = 3  # The size of each action.
+        IMAGE_WIDTH = 64  # The width of each image.
+        IMAGE_HEIGHT = 64  # The height of each image.
+        BUFFER_SIZE = 1024  # The size of the replay buffer.
+        BATCH_SIZE = 128  # Batch size of each sample from the replay buffer.
+        NUM_SAMPLES = 10  # Number of times to sample from the replay buffer.
+
+        replay_buffer = ReplayBuffer(
+            buffer_size=BUFFER_SIZE,
+            batch_size=BATCH_SIZE,
+            observation_type=adapters.AdapterType.DefaultObservationImage,
+            device_name="cpu",
+        )
+
+        (
+            states,
+            next_states,
+            previous_actions,
+            actions,
+            rewards,
+            dones,
+        ) = generate_image_transitions(
+            TRANSITIONS, STACK_SIZE, IMAGE_HEIGHT, IMAGE_WIDTH, ACTION_SIZE
+        )
+
+        for state, next_state, action, previous_action, reward, done in zip(
+            states, next_states, actions, previous_actions, rewards, dones
+        ):
+            replay_buffer.add(
+                state=state,
+                next_state=next_state,
+                action=action,
+                prev_action=previous_action,
+                reward=reward,
+                done=done,
+            )
+
+        for _ in range(NUM_SAMPLES):
+            sample = replay_buffer.sample()
+
+            for state, action, reward, next_state, done, _ in zip(*sample):
+                state = state.numpy()
+                action = action.numpy()
+                reward = reward.numpy()[0]
+                next_state = next_state.numpy()
+                done = True if done.numpy()[0] else False
+
+                index_of_state = None
+                for index, original_state in enumerate(states):
+                    if np.array_equal(original_state, state):
+                        index_of_state = index
+                        break
+
+                self.assertIn(state, states)
+                self.assertIn(next_state, next_states)
+                self.assertIn(action, actions)
+                self.assertIn(reward, rewards)
+                self.assertIn(done, dones)
+                self.assertTrue(np.array_equal(state, states[index_of_state]))
+                self.assertTrue(np.array_equal(next_state, next_states[index_of_state]))
+                self.assertTrue(np.array_equal(action, actions[index_of_state]))
+                self.assertEqual(reward, rewards[index_of_state])
+                self.assertEqual(done, dones[index_of_state])
+
+    # def test_vector_replay_buffer(self):
```

Comment:
Are we going to implement the vector-based replay buffer test later on?

Reply:
Ideally yes, although I think it is a lower priority now with everything else that is going on. We (or someone else) can implement it in the future.